processor/src/main/java/org/mapstruct/ap/internal/model/BeanMappingMethod.java

```diff
@@ -137,14 +137,12 @@ public BeanMappingMethod build() {
                 return null;
             }
 
-            if ( !method.getMappingOptions().isRestrictToDefinedMappings() ) {
```

Comment:
I think that we should always apply `property` and `parameter` name based mappings. The problem with getting the `fish.name` from the `FishTank` should be solved by doing the popping of source reference when we create the forged methods